### DISCUSSION TITLE
Using Default Proxy of User or Envionment (HttpClient.DefaultProxy)

### DISCUSSION THREAD

**Discussion by jwonagel** [jstedfast/MailKit#1812] [Ideas]
All clients (such as SmtpClient, ImapClient, etc.) in MailKit have the option to set a proxy server. The HttpClient and many other libraries use the system default proxy or environment variables (HTTP_PROXY, HTTPS_PROXY, NO_PROXY) to configure the proxy.

It would be beneficial if MailKit also supported a default proxy, similar to HttpClient.DefaultProxy (https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.defaultproxy?view=netcore-3.1). In my opinion, the clients could leverage the HttpClient default proxy, which would simplify the implementation. The feature would only be available on .NET Core 3.0 and newer because DefaultProxy was introduced in .NET Core 3.0.

I would implent it like the following SmtpClient example:

```csharp
    var smtpClient = new SmtpClient();
    smtpClient.UseSystemDefaultProxy = true;
```

I can implement this feature if it's desired and you don't have the time to do it.

**Comment by jstedfast** [comment]
I think it could be useful to have a ProxyClient.DefaultProxy static property that reads environment variables.

I think the nicest approach would be to do:

client.Proxy = ProxyClient.DefaultProxy

**Reply by jwonagel** [reply]
Should I implement it and create a pull request? I should have time this or next week.

**Reply by jstedfast** [reply]
Yea, that would be helpful! Thanks!

**Comment by jwonagel** [comment]
So I could implement a draft that uses the DefaultProxy of the HTTP Client. @jstedfast what do you think about it? If it is ok, I would finalize it and add some unit tests. Heres the link to the diff https://github.com/jstedfast/MailKit/compare/master...jwonagel:MailKit:features/systemproxy?expand=1

**Reply by jstedfast** [reply]
I took a quick look and it looked good, but I had a few questions/suggestions but I wasn't able to comment directly on the diff. I'll try to collect my thoughts and write up some suggestions I had for you (I looked last night and before I could write a response, I got sidetracked and now I don't remember what I was going to say haha).

**Reply by jstedfast** [reply]
In ProxyClient.cs, you added `using Org.BouncyCastle.Security;` - is this needed? Doesn't seem to be.

```csharp
#if NET6_0_OR_GREATER
#nullable enable
		public static IProxyClient DefaultProxy => _systemProxy ??= new SystemProxy ();

		private static SystemProxy? _systemProxy;

#nullable disable
#endif
```

I don't use the `_` naming convention. I also don't use the `=>` syntax (mostly because the codebase has existed since before that feature was added).

In SystemProxy.cs, you are using the newer namespace coding style which is inconsistent with the rest of MailKit :(

I think that's all the issues I had with it (pretty minor).

**Reply by jstedfast** [reply]
I ended up finishing this patch.

**Reply by jwonagel** [reply]
Thank you for finishing it.  I've had a really busy week and though finishing it this week.
